### ISSUE TITLE
``/run/user/1000/`` was owner by root

### ISSUE THREAD

**Issue by SuperSandro2000** [arkane-systems/genie#188]
**Windows version (build number):**
19043.1149

**Linux distribution:**
Ubuntu Hirsute

**Genie version:**
``--version option cannot be combined with other arguments.``

but I installed the package with the version 1.43

**Describe the bug**
``/run/user/1000/`` was owned by root

**If the bug involves `systemctl` or a service running under systemd, confirm that you are running inside the bottle:**
The output of `genie -b`.

**To Reproduce**
1. Enter Genie in one terminal tab ``genie -v -i``
2. Switch to a second terminal and do not enter genie
2. ``/run/user`` is owned by root and ``mktemp -d --tmpdir="/run/user/${UID}"`` fails outside of genie but succeeds inside

**Expected behavior**
That ``/run/user/1000/`` is owned by 1000.

**Screenshots**

**Additional context**
I could fix it with ``sudo chwon 1000:1000 -R /run/user/1000/``

**Comment by cerebrate** [comment]
If I recall correctly, WSLg support didn't land until 21364. While it includes a check to make sure this sort of thing doesn't happen, could you try temporarily removing the `override.conf` under `/lib/systemd/system/user-runtime-dir@.service.d`, restarting WSL completely, and seeing if the problem still occurs?

**Comment by SuperSandro2000** [comment]
> If I recall correctly, WSLg support didn't land until 21364.

Correct.

> could you try temporarily removing the override.conf under /lib/systemd/system/user-runtime-dir@.service.d, restarting WSL completely, and seeing if the problem still occurs?

That fixes the problem.

**Comment by SuperSandro2000** [comment]
Strange enough now I can no longer reproduce the problem and it always just works.

**Comment by PavelSosin-320** [comment]
This is not a bug. Occasionally found the explanation in [Troubleshutting](https://wiki.archlinux.org/title/Systemd/User#How_it_works). Troubleshooting. I hope the explanation is correct for all distros. To create a new user only a distro-specific utility can be used that creates XDG_RUNTIME_DIR correctly. For example for anuser 
ls -al $XDG_RUNTIME_DIR -> drwx------ 5 anuser users 160. It is the critical point if XDG_RUNTIME_DIR is used to store persistent, user-specific data or sockets in a secured manner (SELinux?)

**Comment by PavelSosin-320** [comment]
The related issue appears when attempting to activate Units for non-rootless users: the is no default target and it is impossible to create it in a straightforward manner because of wrong ownership.  Systemd.io offers some procedures [to convert an existing user to a systemd user](https://systemd.io/CONVERTING_TO_HOMED/) but not utility or script. This is not a genie's bug but reference from the documentation can be helpful if somebody wants to start services at rootless user login.

**Comment by SuperSandro2000** [comment]
I am not using systemd-homed. A script is just using the private temp and failing. Fixed it by changing the check to ``[[ -w /run/user/$ID ]]``.

**Comment by cerebrate** [comment]
You know, it only just occurred to me to inquire _why_ you were manually creating the directory in step 2 above.

Because it shouldn't exist after running only `genie -i`. `/usr/run/{UID}` directories are created by pam_systemd.so when a user logs in, and so (a) they are never created by WSL per se, since it doesn't run logins, and (b) wouldn't be created after only `genie -i`, because no-one's logged into the bottle yet.

So you're manually creating it in that step before the normal system mechanisms for doing so ever had a chance to create it. Why are you doing that?

**Comment by PavelSosin-320** [comment]
I'm not sure that [[ -w /rub/user/$UID]] is so reliable. logind may clean resources earlier when last user's session is logged out and not when the user is removed. It has to be checked using loginctl. At least, systemd works with CGs in that way. Another consideration is that temp files MUST be in TMP filesystem. Is it mounted manually too? XDG_RUNTIME_DIR is not for temp files. I suppose that genie -i runs systemd-tmpfiles-setup as a part of systemd boot sequence anyway simply because of the Linux boot sequence  is defined in that way, without being explicitly invoked:
systemctl status systemd-tmpfiles-setup
● systemd-tmpfiles-setup.service - Create Volatile Files and Directories
     Loaded: loaded (/usr/lib/systemd/system/systemd-tmpfiles-setup.service; static)
     Active: active (exited) since Mon 2021-08-09 10:26:06 IDT; 9h ago
       Docs: man:tmpfiles.d(5)
             man:systemd-tmpfiles(8)
    Process: 94 ExecStart=systemd-tmpfiles --create --remove --boot --exclude-prefix=/dev (code=exited, status=0/SUCCESS)
   Main PID: 94 (code=exited, status=0/SUCCESS)

Aug 09 10:26:06 MSI-wsl systemd[1]: Starting Create Volatile Files and Directories...
Aug 09 10:26:06 MSI-wsl systemd[1]: Finished Create Volatile Files and Directories.

**Comment by cerebrate** [comment]
While some subfolders of them are, user runtime directories under /run are _not_ managed by systemd-tmpfiles, as a quick inspection of /lib/tmpfiles.d will demonstrate.

When you log in, _pam_systemd.so_ notifies the system _systemd_ instance (via systemd-logind, which has no other involvement in this process than notifying systemd that a login session has started or stopped) to create the user session/scope (as an instance of the _user@{UID}.service_) and an instance of the _user-runtime-dir@{UID}.service_ (which the _user@{UID}.service_ depends on), which in turn creates the _/run/user/{UID}_ directory and is responsible for deleting it when the system _systemd_ stops the _user-runtime-dir@{UID}.service_ , when the user session terminates. Typically, but not always, on logout.

This is all documented in the _user@.service(5)_ man page.

**Comment by PavelSosin-320** [comment]
@cerebrate I see that user@{UID}.service doesn't create tmp directories but if somebody wants to create manually "temporary" files in the session's scope he can be disappointed because tmp files and file.s created in the session's scope have a different lifetime. RedHat folks build on it when developing applications runnable by rootless users. Who is developing such an application has to understand clearly how the mechanism of the user separation works - errors can cause security breaches or access errors. /run/user/, /tmp/ (mainly used by rootfull services because it is shared) , and /home/user are places for different things.

**Comment by SuperSandro2000** [comment]
> So you're manually creating it in that step before the normal system mechanisms for doing so ever had a chance to create it. Why are you doing that?

I am not manually creating the directory. ``mktemp`` creates a directory in ``/run/user/$UID``.

```ShellSession
$ mktemp -d --tmpdir="/run/user/${UID}"
/run/user/1000/tmp.kraM3Oupsa
```

> XDG_RUNTIME_DIR is not for temp files.

Surprisingly many programs use it for a private temp directory and at least on a normal Linux system it is a tmpfs.

**Comment by PavelSosin-320** [comment]
@SuperSandro2000 Maybe, it will help you: I'm login to my Fedora distro as non-root user using Windows Terminal profile
            {
                "commandline": "wsl -d fedoraremix -u root genie -l",
                "guid": "{9bf73832-84d5-44d5-87d9-3e60f3cd930e}",
                "hidden": false,
                "name": "FedoraPavel"
            }
So, the pam is always called as root and then switches to the login user.
It is also very convenient because I can quickly switch users and work as more than one user simultaneously opening and switching WT tabs. In my scenario /run/user/{$UID} folder has the correct owner and access rights. /home/user-name created by WSL still has wrong access rights but systemd.io expects this scenario.

**Comment by cerebrate** [comment]
> I am not manually creating the directory. mktemp creates a directory in /run/user/$UID.

My mistake; it's been a long damn week. But the point of that post remains: you don't appear to be doing anything in between steps 1 and 2 of your repro case (i.e., entering the bottle and thus creating a user session) that would cause that directory to be created, which would be why it isn't.

Do that and get back to me.

**Comment by PavelSosin-320** [comment]
Either TMPDIR env variable is missed or bug in mktemp. Indeed both /tmp and TMPDIR are missed because it is unpredictable in WSL.  Maybe $TMPDIR in the distro -> /mnt/c/temp in Windows like in Docker Desktop ?

**Comment by cerebrate** [comment]
Please stop helping.

Until we resolve the very basic "not actually being inside the bottle and therefore never initializing the runtime directory" question, any further speculation, especially in areas at best tangential to this issue, is useless.

**Comment by SuperSandro2000** [comment]
> you don't appear to be doing anything in between steps 1 and 2 of your repro case (i.e., entering the bottle and thus creating a user session) that would cause that directory to be created, which would be why it isn't.

I have updated the reproducer.

```
Steps to reproduce the behavior:
1. Enter Genie in one terminal tab ``genie -v -i``
2. Switch to a second terminal and do not enter genie
2. ``/run/user`` is owned by root and ``mktemp -d --tmpdir="/run/user/${UID}"`` fails outside of genie but succeeds inside
```

I forgot to mention that this is done by my prompt so I don't really need to do anything but of course you can't know that.

**Comment by PavelSosin-320** [comment]
Neither genie, mktemp nor users are guilty. The full explanation is given in  [Using /run/user as temp](https://access.redhat.com/solutions/2060373). It also has a limited size. Pre-setting of XDG_RUNTIME_DIR will not help because it is managed by pam service that runs every login attempt that, in turn, runs only inside the bottle.
Once logind is started during initialization, i.e. genie -i or genie -s developer has no reasons to care or check something.

**Comment by cerebrate** [comment]
To repeat both the documentation and my comment above:

**You are NOT entering the bottle. `genie -i` DOES NOT DO THAT.**

I quote:

> **NOTE 2:** genie -i DOES NOT enter the bottle for you. It is important to remember that the genie bottle functions like a container, with its own cgroups and separate pid and mount namespaces. While some systemd or systemd-service powered things may work when invoked from outside the bottle, this is ENTIRELY BY CHANCE, and is NOT A SUPPORTED SCENARIO. You must enter the bottle using `genie -s`, `genie -l` or `genie -c` first.

tl;dr: your repro case shows you doing something unsupported and documented as unlikely to work. Therefore, it not working is not an issue.

**Comment by PavelSosin-320** [comment]
I do enter the bottle when I do genie -l ( see above). I start distro using genie -s because I need to ensure that the Networking layer, logind, and other root-only units start before my own units in ~/.config/user. Then I can safely start as rootless user. My own rootless systemd instance is created and all my services: Theia, Podman,vscodeserver are managed using systemctl --user, other systemd CTLs are working as --user and everything is completely OK! I have to monitor CGs structure using systemd-cgls because I'm solving some Podman issue via managing Podman REST API server as unit via Unit dependency from Editor backend container.

**Comment by PavelSosin-320** [comment]
@cerebrate I'm in favor to close this issue too! If the user works correctly and understands what happens behind the facade of Linux Desktop everything works perfectly! I need to start the distro with wsl genie -s under the root user to execute the regular boot sequence using systemd. The opened window is inside the bottle, correct? Then if I see is-system-running returns running, otherwise - investigate. At this point I have the system is initialized, init, system, and user slices are created. The opened WS window is my "admin" console, nothing else, like in the real Linux desktop. If the Networking stack failed to start then the root user and all rootless users as well will have networking problems. Then I open new sessions using wsl genie -l and log on as user using the new WT window. If logon has succeeded I'm inside bottle, correct? If systemctl --user is-system-running returns running I'm inside the bottle and inside settled session and can expect that environment is settled, tmp files created, the session has reached the default target and now I can work with standard Fedora utilities and CLI's, start my services by Units using the opened window without any care, without checking genie -b. Everything is done via the regular, trusted systemd mechanism. If I'm not inside the genie bottle I have no chance to reach the desired targets. The separation between root user and rootless users namespaces is not only expected but desired systemd V2 feature. The separation of mount namespaces is not relevant if the user uses only mount.fuse utility and the current WSL Kernel with FUSE support. The separation of IPC namespaces is the only side effect but possibly desired by RedHat. If something is wrong from the genie point of view please, describe the exact scenario and I will test.

**Comment by SuperSandro2000** [comment]
> You are NOT entering the bottle. genie -i DOES NOT DO THAT.

I am entering the bottle in one terminal but I just copied the wrong command in my giant dotfiles.

But I think I understand now that this is just a bad combination of things to do.
I either check if /run/user/$UID is writtable or change my dotfiles that I always enter the bottle and remove the timeout to abort entering it.

**Comment by PavelSosin-320** [comment]
@cerebrate By the way - genie -b question has different answers for the root and every rootless user session. When users root and dog can be inside their bottle, user cat can be outside. Fortunately, systemctl [ --user | --system ] always gives the right answer.
